presto-rcfile/src/main/java/com/facebook/presto/rcfile/RcFileWriter.java

```diff
@@ -0,0 +1,407 @@
+
+        this.output = output;
+
+        // write header
+        output.writeBytes(RCFILE_MAGIC);
+        output.writeByte(CURRENT_VERSION);
+
+        // write codec information
+        output.writeBoolean(codecName.isPresent());
+        if (codecName.isPresent()) {
+            writeLengthPrefixedString(output, utf8Slice(codecName.get()));
+        }
+
+        // write metadata
+        output.writeInt(Integer.reverseBytes(metadata.size() + 1));
+        writeLengthPrefixedString(output, utf8Slice(COLUMN_COUNT_METADATA_KEY));
+        writeLengthPrefixedString(output, utf8Slice(Integer.toString(types.size())));
+        for (Entry<String, String> entry : metadata.entrySet()) {
+            writeLengthPrefixedString(output, utf8Slice(entry.getKey()));
+            writeLengthPrefixedString(output, utf8Slice(entry.getValue()));
+        }
+
+        // write sync sequence
+        output.writeLong(syncFirst);
+        output.writeLong(syncSecond);
+
+        // initialize columns
+        RcFileCompressor compressor = codecName.map(codecFactory::createCompressor).orElse(new NoneCompressor());
+        keySectionOutput = compressor.createCompressedSliceOutput((int) MIN_BUFFER_SIZE.toBytes(), (int) MAX_BUFFER_SIZE.toBytes());
+        columnEncoders = new ColumnEncoder[types.size()];
+        for (int columnIndex = 0; columnIndex < types.size(); columnIndex++) {
+            Type type = types.get(columnIndex);
+            ColumnEncoding columnEncoding = encoding.getEncoding(type);
+            columnEncoders[columnIndex] = new ColumnEncoder(columnEncoding, compressor);
+        }
+        this.minRowGroupSize = (int) minRowGroupSize.toBytes();
+        this.maxRowGroupSize = (int) maxRowGroupSize.toBytes();
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        writeRowGroup();
+        output.close();
+        keySectionOutput.destroy();
+        for (ColumnEncoder columnEncoder : columnEncoders) {
+            columnEncoder.destroy();
+        }
+    }
+
+    public long getRetainedSizeInBytes()
+    {
+        long retailedSize = 0;
+        retailedSize += output.getRetainedSize();
+        retailedSize += keySectionOutput.getRetainedSize();
+        for (ColumnEncoder columnEncoder : columnEncoders) {
+            retailedSize += columnEncoder.getRetailedSizeInBytes();
+        }
+        return retailedSize;
+    }
+
+    public void write(Page page)
+            throws IOException
+    {
+        if (page.getPositionCount() == 0) {
+            return;
+        }
+
+        long pageSize = page.getSizeInBytes();
+        if (pageSize <= maxRowGroupSize || page.getPositionCount() == 1) {
+            bufferPage(page);
+            return;
+        }
+
+        // recursively split page until it is less than the max row group size
+        int positionCount = page.getPositionCount();
```

Comment:
To clarify, what I meant here is to add sanity check here that the sum of the size of the two new pages are the same of the original page. The idea is that this will avoid writing unreasonably encoded files in the event that Block.getSizeInBytes is broken. @dain argued that we should really just add additional tests for Blocks instead.

I'm ok with passing on this comment.